packages/sanity/src/core/form/store/formState.ts

```diff
@@ -128,937 +191,1106 @@ function isChangedValue(value: any, comparisonValue: any) {
-    readOnly,
-    schemaType: props.schemaType,
-    focused: isEqual(props.path, props.focusPath),
-    focusPath: trimChildPath(props.path, props.focusPath),
-    path: props.path,
-    id: toString(props.path),
-    level: props.level,
-    validation,
-    presence,
-    members,
-  }
-}
+  const preparePrimitiveInputState = memoizePreparePrimitiveInputState(
+    function _preparePrimitiveInputState(props) {
+      const filteredPresence = props.presence.filter((item) => isEqual(item.path, props.path))
+      const presence = filteredPresence.length ? filteredPresence : EMPTY_ARRAY
 
-/*
- * Takes a field in context of a parent object and returns prepared props for it
- */
-function prepareArrayOfObjectsMember(props: {
-  arrayItem: {_key: string}
-  parent: RawState<ArraySchemaType, unknown>
-  index: number
-}): ArrayOfObjectsMember {
-  const {arrayItem, parent, index} = props
-
-  const itemType = getItemType(parent.schemaType, arrayItem) as ObjectSchemaType
-
-  const key = arrayItem._key
-
-  if (!itemType) {
-    const itemTypeName = resolveTypeName(arrayItem)
-    return {
-      kind: 'error',
-      key,
-      index,
-      error: {
-        type: 'INVALID_ITEM_TYPE',
-        resolvedValueType: itemTypeName,
-        value: arrayItem,
-        validTypes: parent.schemaType.of,
-      },
-    }
-  }
-
-  const itemPath = pathFor([...parent.path, {_key: key}])
-  const itemLevel = parent.level + 1
+      const validation = props.validation
+        .filter((item) => isEqual(item.path, props.path))
+        .map((v) => ({level: v.level, message: v.message, path: v.path}))
+      return {
+        schemaType: props.schemaType,
+        changed: isChangedValue(props.value, props.comparisonValue),
+        value: props.value,
+        level: props.level,
+        id: toString(props.path),
+        readOnly: props.readOnly === true || props.readOnly?.value,
+        focused: isEqual(props.path, props.focusPath),
+        path: props.path,
+        presence,
+        validation,
+      } as PrimitiveFormNode
+    },
+  )
 
-  const conditionalPropertyContext = {
-    value: parent.value,
-    parent: props.parent,
-    document: parent.document,
-    currentUser: parent.currentUser,
-  }
-  const readOnly =
-    parent.readOnly ||
-    resolveConditionalProperty(parent.schemaType.readOnly, conditionalPropertyContext)
-
-  const fieldGroupState = parent.fieldGroupState?.children?.[key]
-  const scopedCollapsedPaths = parent.collapsedPaths?.children?.[key]
-  const scopedCollapsedFieldsets = parent.collapsedFieldSets?.children?.[key]
-  const comparisonValue =
-    (Array.isArray(parent.comparisonValue) &&
-      parent.comparisonValue.find((i) => i._key === arrayItem._key)) ||
-    undefined
-
-  const itemState = prepareObjectInputState(
-    {
-      schemaType: itemType,
-      level: itemLevel,
-      document: parent.document,
-      value: arrayItem,
+  return Object.assign(
+    function prepareFormState({
```

Comment:
Assigning properties to a function like this feels a little bit peculiar to me – any drawbacks to keeping `prepareFormState` as an individual property on the returned object here?

Reply:
My thought process here is that the rest of the function inside of `prepareFormState` are internal. The only reason i'm exposing them is to wrap them with `jest.fn` so i can call `expect(/* .. .*/).toHaveBeenCalledTimes(/* ... */)`.  I think i might prefix these with an underscore of something to denote that.